### ISSUE TITLE
support nested emphasis delimeters

### ISSUE THREAD

**Issue by brainchild0** [marktext/marktext#2086]
Markdown processors vary in handling of nested emphasis delimiters, but most have some kind of support. 

For example, consider the following text: 

```
**a **b** c**
*a *b* c*
```

Some engines would render the *b* letters specially because of the nesting, but almost all will recognize that all the asterisk characters in this particular example are in fact delimiters.  Basic handling would treat the entire first line as strong emphasis text and the entire second line as emphasis text.  That is, the lines will be rendered as:

> **a **b** c**
> *a *b* c*

I recommend that Mark Text include such support, which it currently lacks, as seen in the below window capture.

![Screenshot from 2020-04-30 09-37-09](https://user-images.githubusercontent.com/34869175/80717030-60c67d00-8ac6-11ea-8d0e-b56cc2ee8dd3.png)

**Comment by fxha** [comment]
Please go ahead and file an issue [here](https://github.com/markedjs/marked). I think the core issue is that both `*a ... c*` and `*b*` are valid options but `*a *b` and `b* c*` aren't because there are spaces between the delimiter and the character.

**Comment by brainchild0** [comment]
If Mark Text uses Marked internally, which I did not know, then I agree that the  issue should be passed upstream.  However,   based on reviewing the issue tracker for Marked, achieving CommonMark compliance is an ongoing activity.

I am reluctant to create or even to mention an issue in Marked, to avoid creating unwanted clutter in their system.

However, since Mark Text is affected, perhaps we can leave open this issue to track progress.

**Comment by fxha** [comment]
> If Mark Text uses Marked internally, which I did not know, then I agree that the issue should be passed upstream.

Yes, we're using a customized fork of marked that is really close to upstream. Please file an issue there because this is a special case and need to be discussed whether and how it should be fixed - both CommonMark and GFM render this a bold/italic.

See [demo](https://marked.js.org/demo/).

**Comment by brainchild0** [comment]
> both CommonMark and GFM render this a bold/italic

In CommonMark it's double emphasis:

```
$ echo '*a *b* c*' | pandoc --from commonmark 
<p><em>a <em>b</em> c</em></p>
$ echo '**a **b** c**' | pandoc --from commonmark 
<p><strong>a <strong>b</strong> c</strong></p>
```

This is described (not terribly well) in the spec.

Here in Github it's also not bold-italic:

Code | Rendering
--- | ---
`*a *b* c*` | *a *b* c*
`**a **b** c**` | **a **b** c**



The Marked group is working to gain compliance with the standards test suites.   Once they  do, then the issue would resolve. They already have an issue to track this progress.

**Comment by fxha** [comment]
Failed GFM examples:

- 382
- 416
- 427
- ...

This might be an easy fix because `_foo __bar__ baz_` is parsed correctly but `_foo _bar_ baz_` not.

**Comment by brainchild0** [comment]
If we collect a small set of compliance problems that are high-impact for the project, then I believe it would be reasonable to communicate to the Marked group, in case it wants to prioritize those issues. 

However, simply reminding the group that the project is not fully compliant yet would seem unhelpful.

**Comment by brainchild0** [comment]
Apparently upstream gifts may be coming  sooner than expected.  

See https://github.com/markedjs/marked/pull/1686.

**Comment by brainchild0** [comment]
Improved support for nested emphasis is now available in marked. PR  #2277 attempts to integrate the new version, by changing `package.json` but is not currently successful. I am not aware of the customized version of marked mentioned earlier.

**Comment by fxha** [comment]
@brainchild0 We're using a custom version of marked.js and we need to rebase our custom [marked.js](https://github.com/marktext/marktext/tree/develop/src/muya/lib/parser/marked). The package included in the project is to test our version against the upstream one.

**Comment by brainchild0** [comment]
@fxha: It seems that the fork is heavily restructured, into multiple files. I find no obvious way to automate the rebasing from the downstream source.

I'm sure that you encountered various constraints writing the fork, but I am wondering about moving to some kind of more automatic and reliable maintenance process. Is it possible to consider including the source in a submodule, and then modifying the library through a static patch applied by a build preprocessor? Or perhaps some of the additional functionality may be gained through [monkey patching](https://en.wikipedia.org/wiki/Monkey_patch)? Even a forked repository that can use Git merging may help improve maintenance in the future.

**Comment by fxha** [comment]
This seems to be an issue with our [frontend](https://github.com/marktext/marktext/blob/2176d241594397df345a408a9f819f01f469fbfa/src/muya/lib/parser/rules.js#L14) and not marked.js.
